akka-projection-slick/src/main/scala/akka/projection/slick/OffsetStore.scala

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (C) 2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.projection.slick
+
+import java.util.UUID
+
+import akka.Done
+import akka.annotation.ApiMayChange
+import slick.jdbc.JdbcProfile
+import akka.persistence.query
+
+import scala.concurrent.{ ExecutionContext, Future }
+
+@ApiMayChange
+class OffsetStore[P <: JdbcProfile](db: P#Backend#Database, profile: P) {
+
+  import profile.api._
+
+  // offset manifest
+  private val StringM = "STR"
+  private val LongM = "LNG"
+  private val IntM = "INT"
+  private val SequenceM = "SEQ"
+  private val TimeBasedUUIDM = "TBU"
+
+  def readOffset[Offset](projectionId: String)(implicit ec: ExecutionContext): Future[Option[Offset]] = {
+    val action =
+      offsetTable.filter(_.projectionId === projectionId).result.headOption.map(row => fromRowToOffset[Offset](row))
+
+    db.run(action)
+  }
+
+  def saveOffset[Offset](projectionId: String, offset: Offset)(implicit ec: ExecutionContext): slick.dbio.DBIO[Done] = {
+    val (offsetStr, manifest) = stringOffsetAndManifest(offset)
+    offsetTable.insertOrUpdate(OffsetRow(projectionId, offsetStr, manifest)).map(_ => Done)
+  }
+
+  /**
+   * Deserialize an offset from a result row.
+   * The offset is converted from its string representation to its real time.
+   */
+  private def fromRowToOffset[Offset](row: Option[OffsetRow]): Option[Offset] = {
+    row match {
+      case Some(OffsetRow(_, offset, StringM))   => Some(offset.asInstanceOf[Offset])
+      case Some(OffsetRow(_, offset, LongM))     => Some(offset.toLong.asInstanceOf[Offset])
+      case Some(OffsetRow(_, offset, IntM))      => Some(offset.toInt.asInstanceOf[Offset])
+      case Some(OffsetRow(_, offset, SequenceM)) => Some(query.Sequence(offset.toLong).asInstanceOf[Offset])
+      case Some(OffsetRow(_, offset, TimeBasedUUIDM)) =>
+        Some(query.TimeBasedUUID(UUID.fromString(offset)).asInstanceOf[Offset])
+      case Some(OffsetRow(id, offset, manifest)) =>
+        throw new IllegalStateException(s"Unknown offset manifest [$manifest] for projection id [$id]")
+      case None => None
+    }
+  }
+
+  /**
+   * Convert the offset to a tuple (String, String) where the first element is
+   * the String representation of the offset and the second its manifest
+   */
+  private def stringOffsetAndManifest[Offset](offset: Offset): (String, String) = {
+    offset match {
+      case s: String                => s -> StringM
+      case l: Long                  => l.toString -> LongM
+      case i: Int                   => i.toString -> IntM
+      case seq: query.Sequence      => seq.value.toString -> SequenceM
+      case tbu: query.TimeBasedUUID => tbu.value.toString -> TimeBasedUUIDM
+      case _                        => throw new IllegalArgumentException(s"Unsupported offset type, found [${offset.getClass}]")
+    }
+  }
+
+  private class OffsetStoreTable(tag: Tag) extends Table[OffsetRow](tag, "PROJECTION_OFFSET_STORE") {
```

Comment:
I haven't used slick in a long time, but it seems that the `tag` is not encoded into the table name. Since the table name is hard coded it will be in contention by different projections.

Reply:
Ah, I see `PROJECTION_ID` is a field in the table.

Reply:
yep, that projection_id will be the `name` + `key`. I have introduced a `ProjectionId` type for that. Pushing soon...